### ISSUE TITLE
Memory leak when using threads

### ISSUE THREAD

**Issue by toots** [ocaml/ocaml#11289]
**This bug has been tested with OCaml `4.13.0`. I am in the process of testing it with `5.0.0` (on a m1 mac)**

The following program:
```ocaml
let rec f () =
  let th = Thread.create (fun () ->
    let b = Bytes.make (10*4096) ' ' in
    Printf.printf "New Thread with %d bytes\n%!" (Bytes.length b);
    Unix.sleepf 0.1) ()
  in
  Thread.join th;
  f ()

let () = f ()
```

Seems to leave the memory allocated in the thread for too long. In my test, the program quickly rose to ~60Mo of RAM within minutes:
![mem_usage](https://user-images.githubusercontent.com/871060/171551669-4ac0c3a6-80f1-4841-bcd3-a544cb8a7e17.png)

The same program with a call to `full_major` does not exhibit the issue:
```ocaml
let rec f () =
  let th = Thread.create (fun () ->
    let b = Bytes.make (10*4096) ' ' in
    Printf.printf "New Thread with %d bytes\n%!" (Bytes.length b);
    Unix.sleepf 0.1) ()
  in
  Thread.join th;
  Gc.full_major();
  f ()

let () = f ()
```
![mem_usage_full_major](https://user-images.githubusercontent.com/871060/171551738-cfbd2a8e-8eda-4063-b633-ec01e1048045.png)

This seems to indicate that the GC is underestimating the amount of memory waiting to be cleaned up from the thread's allocations.

**Comment by nojb** [comment]
I tried with 4.14.0 and it seems to work well, in the sense that the GC runs periodically and memory use never goes over ~12MB.

**Comment by gasche** [comment]
It's not clear to me that there is a bug anyway: if there is enough memory available on the system, isn't it reasonable for the GC to waste memory to reduce its working time? I guess we would prefer constant-live-memory program to "naturally" exhibit constant memory usage (that is: even in unconstrained environments), and this depends on the GC tuning heuristics -- which I guess have changed a bit between 4.13 and 4.14, and changed quite a bit again between 4.14 and 5.0-beta.

**Comment by toots** [comment]
I can definitely reproduce with `4.14.0` here. I'll let it run for a while and will report.

**Comment by toots** [comment]
> It's not clear to me that there is a bug anyway: if there is enough memory available on the system, isn't it reasonable for the GC to waste memory to reduce its working time? I guess we would prefer constant-live-memory program to "naturally" exhibit constant memory usage (that is: even in unconstrained environments), and this depends on the GC tuning heuristics -- which I guess have changed a bit between 4.13 and 4.14, and changed quite a bit again between 4.14 and 5.0-beta.

I think a program ought to use a "reasonable" amount of memory. This is particularly true in virtualized contexts. I understand that the nuance here on GC CPU usage but, at the very least, the memory usage I am seeing is not what one would expect for such a small predictable program.

**Comment by toots** [comment]
After one night running:
<img width="284" alt="Screen Shot 2022-06-02 at 9 03 23 AM" src="https://user-images.githubusercontent.com/871060/171647339-b7ab4c26-8cf9-4238-8e42-7f9f3aae516a.png">

**Comment by nojb** [comment]
Strange, in my machine (but I didn't let it run very long), memory use maxes out at ~10Mo (click to see):

![demo](https://user-images.githubusercontent.com/113560/171649095-5ee34101-a91e-4dcb-aad9-2ffd2faf88f1.gif)

**Comment by toots** [comment]
@nojb I've been able to reproduce with 4.13 and 4.14 on two arm machines, one macos and one m1. I believe that @smimram has been able to reproduce on amd64.

Using `mem_usage`, this script will print out its growing memory usage and can be plotted:
```ocaml
let rec f () =
  let th = Thread.create (fun () ->
    let b = Bytes.make (10*4096) ' ' in
    Unix.sleepf 0.1;
    b) ()
  in
  Thread.join th;
  Printf.printf "%d\n%!" Mem_usage.((info ()).process_physical_memory);
  f ()

let () = f ()
```

**Comment by nojb** [comment]
I cannot reproduce on my side, mem usage never goes above 10/12 Mo. @gasche can you give it a try if you have time?

**Comment by toots** [comment]
I tried to reproduce in a `docker` environment and it was stable. Steps:
```shell
% docker run -ti --entrypoint /bin/bash debian:stable
# apt-get update
# apt-get install opam
# opam init --comp=4.14.0 --disable-sandbox
# eval $(opam env)
# opam install ocamlfind mem_usage
(...add test script...)
# ocamlfind ocamlopt -linkpkg -package unix -package threads -package mem_usage -thread ./test.ml -o test
# ./test
```

**Comment by toots** [comment]
Ubuntu jammy on `docker` with `4.13` does definitely leak:
![ubuntu-mem-leak](https://user-images.githubusercontent.com/871060/171905241-29a3d46b-0b3c-4934-9b3c-a2994b2c7661.png)

Testing with 4.14 on the same distribution next.

**Comment by xavierleroy** [comment]
Guys: instead of arguing "it's reproducible!" "no it's not!", could we please 
1- agree that yes it is reproducible on Linux but perhaps not on Windows (pointing to a possible source for the issue), and
2- start debugging the problem?

If I see one more pasted graphics in this issue report I'm going to scream.

**Comment by toots** [comment]
> Guys: instead of arguing "it's reproducible!" "no it's not!", could we please 1- agree that yes it is reproducible on Linux but perhaps not on Windows (pointing to a possible source for the issue), and 2- start debugging the problem?
> 
> If I see one more pasted graphics in this issue report I'm going to scream.

I hear you. I'm just not sure if it's current personally, I feel like a lot of things have changed with the GC re: threads lately. I'll try to avoid graphs and will just report yes/no 🙂

**Comment by toots** [comment]
`4.14` appears to be stable. I'm gonna test for `5.00` next.

**Comment by toots** [comment]
I will confirm once the test has ran for a longer time but it does appear that the example also leaks with OCaml 5.

**Comment by toots** [comment]
Ok y'all, I have a better report. I felt that graphs would help visualize what's going on, hope this is okay @xavierleroy.

### tl;dr:
* 4:13: **present**
* 4.14: **not present**
* 5: **present**

### Reproduction script:
```ocaml
let print_interval = 10.
let start_time = Unix.time ()
let last_print = ref 0.

let rec f () =
  let th =
    Thread.create
      (fun () ->
        let b = Bytes.make (10 * 4096) ' ' in
        Unix.sleepf 0.1;
(* Uncomment when testing with GC
        Gc.full_major ();
*)
        b)
      ()
  in
  Thread.join th;
  let current_time = Unix.time () in
  if !last_print +. print_interval < current_time then (
    last_print := current_time;
    let info = Mem_usage.info () in
    Printf.printf
      {|{"elapsed_time": %f, "total_virtual_memory": %d, "total_physical_memory": %d, "total_used_virtual_memory": %d, "total_used_physical_memory": %d, "process_virtual_memory": %d, "process_physical_memory": %d}
%!|}
      (current_time -. start_time)
      info.Mem_usage.total_virtual_memory info.Mem_usage.total_physical_memory
      info.Mem_usage.total_used_virtual_memory
      info.Mem_usage.total_used_physical_memory
      info.Mem_usage.process_virtual_memory
      info.Mem_usage.process_physical_memory);
  f ()

let () = f ()
```
Script will print all memory info every 10s.

### Graphs without `Gc.full_major()`:
* **4.13.1**
![mem-413](https://user-images.githubusercontent.com/871060/174461002-6063fee1-1f5d-4856-8a1c-5ba37926aeb5.png)
This one grew pretty quickly!

* **4.14.0**
![mem-414](https://user-images.githubusercontent.com/871060/174461006-fe13ef49-57f2-4144-913b-0b9524c69362.png)

This one seems stable. Some intense CPU operation started in the middle, explaining the change in dynamic

* **5.00**
![mem-5](https://user-images.githubusercontent.com/871060/174461018-787ef5c8-50ad-44fa-9d4e-5531056b6039.png)

Slow creeping. Same thing with CPU usage in the middle.

### Graphs with `Gc.full_major()`:

I didn't let them run as long, I don't think that was needed, memory should be stable right away.

* **4.13.1**

![mem-413](https://user-images.githubusercontent.com/871060/174461047-b40fbac7-267d-4547-828d-7d88b1ebc9a3.png)

Same creeping

* **4.14.0**
![mem-414](https://user-images.githubusercontent.com/871060/174461049-4e91ccae-5856-47c5-b2ff-d436d256ed1c.png)

Steady as expected!

* **5.00**
![mem-5](https://user-images.githubusercontent.com/871060/174461054-36bf980a-21f1-4cb8-98cd-a44ffb91fec6.png)

Slow creep.

**Comment by toots** [comment]
I've been running some experiments using `-fsanitize=address` (https://github.com/ocaml/opam-repository/pull/21667) and this shows up for each thread creation cycle:
```
==27144==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 40400 byte(s) in 101 object(s) allocated from:
    #0 0xffff8d9a1540 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0xaaaad0aa1998 in caml_stat_alloc runtime/memory.c:542
    #2 0xaaaad0a84428 in caml_maybe_expand_stack runtime/fiber.c:313
    #3 0xaaaad0a767f0 in caml_callback_exn runtime/callback.c:154
    #4 0xaaaad0a52420 in caml_thread_start /home/opam/.opam/5.1/.opam-switch/build/ocaml-variants.5.0.0+trunk/otherlibs/systhreads/st_stubs.c:493
    #5 0xffff8d877620 in start_thread (/lib/aarch64-linux-gnu/libpthread.so.0+0x7620)
    #6 0xffff8d723498  (/lib/aarch64-linux-gnu/libc.so.6+0xd1498)
```

**Comment by gasche** [comment]
I think I see an issue indeed in the threads code.

```c
/* Create and setup a new thread info block.
   This block has no associated thread descriptor and
   is not inserted in the list of threads. */

static caml_thread_t caml_thread_new_info(void)
{
  caml_thread_t th;
  caml_domain_state *domain_state;
  uintnat stack_wsize = caml_get_init_stack_wsize();

  domain_state = Caml_state;
  th = NULL;
  th = (caml_thread_t)caml_stat_alloc_noexc(sizeof(struct caml_thread_struct));
  if (th == NULL) return NULL;
  th->descr = Val_unit;
  th->next = NULL;
  th->prev = NULL;
  th->domain_id = domain_state->id;
  th->current_stack = caml_alloc_main_stack(stack_wsize);
  if (th->current_stack == NULL) {
    caml_stat_free(th);
    return NULL;
  }
  th->c_stack = NULL;
  th->local_roots = NULL;
  th->backtrace_pos = 0;
  th->backtrace_buffer = NULL;
  th->backtrace_last_exn = Val_unit;
  th->gc_regs = NULL;
  th->gc_regs_buckets = NULL;
  th->exn_handler = NULL;

#ifndef NATIVE_CODE
  th->trap_sp_off = 1;
  th->trap_barrier_off = 2;
  th->external_raise = NULL;
#endif

  return th;
}

[...]

/* Remove a thread info block from the list of threads.
   Free it and its stack resources. */

static void caml_thread_remove_info(caml_thread_t th)
{
  if (th->next == th)
    All_threads = NULL; /* last OCaml thread exiting */
  else if (All_threads == th)
    All_threads = th->next;     /* PR#5295 */
  th->next->prev = th->prev;
  th->prev->next = th->next;
  caml_free_stack(th->current_stack);
  caml_stat_free(th);
  return;
}
```

The problem is that, over the lifetime of the thread, some of the values initialized with `NULL` at thread-creation time may have been turned into pointers to malloc-ed memory, and those blocks are not freed in the symmetric `remove_info` function -- except for the explicit call to `caml_free_stack`, which is the only attempt to collect thread-local resources.

These resources are allocated on-demand by the runtime, so their allocation code is often not exposed in the corresponding .h files, and sometimes there even isn't any deallocation code. When they are domain-local, it's not a big deal as they get reused on the next domain creation (with our current approach of reusing domain_state slots), but threads have shorter lifetimes and may be created in large numbers so this is a problem.

I haven't done a full audit but I see an issue for at least:
- `gc_regs_bucket`, which is the LeakSanitizer trace you are pointing at
- `backtrace_buffer`

Cleaning this up will require some refactoring.

cc @Engil ; should we try working on this together?

**Comment by gasche** [comment]
Note: I'm making the assumption that it is possible for threads to be collected while they are not active -- the resources are "saved" in the `caml_thread_t` struct, then collected, without being moved into `Caml_state` again. If threads were always "resumed" on termination, there would be no leak here. But clearly at least the call stack is explicitly collected, so I assume that such abrupt termination is possible.

**Comment by gasche** [comment]
I'm looking at this again with @Engil and @fabbing and we changed the test case to not use @toots' [mem_usage](https://github.com/savonet/ocaml-mem_usage) library. The new code is as follows:

```ocaml
let print_interval = 10.
let start_time = Unix.time ()
let last_print = ref 0.

let rec f () =
  let th =
    Thread.create
      (fun () ->
        let b = Bytes.make (10 * 4096) ' ' in
        Unix.sleepf 0.1;
(*
(* Uncomment when testing with GC *)
        Gc.full_major ();
*)
        b)
      ()
  in
  Thread.join th;
  let current_time = Unix.time () in
  let pid = Unix.getpid () in
  if !last_print +. print_interval < current_time then (
    Printf.ksprintf (fun cmd -> ignore (Sys.command cmd))
      "grep -i rss /proc/%d/status" pid
  );
  f ()

let () = f ()
```

We are currently not sure whether this version leaks (on trunk, on my linux machine).

**Comment by gasche** [comment]
On my machine the test program does not leak very clearly as-is, but adding `try failwith "" with _ -> ()` inside the thread, which forces to allocate a new backtrace buffer, leaks very clearly. All sources of leaks we could identify by following [my hypothesis above](https://github.com/ocaml/ocaml/issues/11289#issuecomment-1168402610) should be fixed in #11406. There may of course be other leaks remaining in the runtime.

**Comment by toots** [comment]
I believe that this was fixed by: https://github.com/ocaml/ocaml/pull/11406
